RecoVertex/BeamSpotProducer/plugins/BeamSpotCompatibilityChecker.cc

```diff
@@ -0,0 +1,242 @@
+  static constexpr int cmToum = 10000;
+  const double warningThreshold_;
+  const double throwingThreshold_;
+  const bool verbose_;
+  const bool dbFromEvent_;
+  const edm::EDGetTokenT<reco::BeamSpot> bsFromEventToken_;             // beamSpot from the event
+  edm::ESGetToken<BeamSpotObjects, BeamSpotObjectsRcd> bsFromDBToken_;  // beamSpot from the DB
+  edm::EDGetTokenT<reco::BeamSpot> dbBSfromEventToken_;                 // beamSpot from the DB (via event)
+};
+
+//
+// constructors and destructor
+//
+BeamSpotCompatibilityChecker::BeamSpotCompatibilityChecker(const edm::ParameterSet& iConfig)
+    : warningThreshold_(iConfig.getParameter<double>("warningThr")),
+      throwingThreshold_(iConfig.getParameter<double>("errorThr")),
+      verbose_(iConfig.getUntrackedParameter<bool>("verbose", false)),
+      dbFromEvent_(iConfig.getParameter<bool>("dbFromEvent")),
+      bsFromEventToken_(consumes<reco::BeamSpot>(iConfig.getParameter<edm::InputTag>("bsFromEvent"))) {
+  //now do what ever initialization is needed
+  if (warningThreshold_ > throwingThreshold_) {
+    throw cms::Exception("ConfigurationError")
+        << __PRETTY_FUNCTION__ << "\n Warning threshold (" << warningThreshold_
+        << ") cannot be smaller than the throwing threshold (" << throwingThreshold_ << ")" << std::endl;
+  }
+  if (dbFromEvent_) {
+    edm::LogWarning("BeamSpotCompatibilityChecker")
+        << "!!!! Warning !!!\nThe Database Beam Spot is going to be taken from the Event via BeamSpotProducer!";
+    dbBSfromEventToken_ = consumes<reco::BeamSpot>(iConfig.getParameter<edm::InputTag>("bsFromDB"));
+  } else {
+    bsFromDBToken_ = esConsumes<BeamSpotObjects, BeamSpotObjectsRcd>();
+  }
+}
+
+//
+// member functions
+//
+
+// ------------ method called for each event  ------------
+void BeamSpotCompatibilityChecker::analyze(edm::StreamID sid,
+                                           edm::Event const& iEvent,
+                                           edm::EventSetup const& iSetup) const {
+  using namespace edm;
+  reco::BeamSpot spotEvent, spotDB;
+
+  edm::Handle<reco::BeamSpot> beamSpotFromEventHandle;
+  iEvent.getByToken(bsFromEventToken_, beamSpotFromEventHandle);
+  spotEvent = *beamSpotFromEventHandle;
+
+  double evt_BSx0_ = spotEvent.x0();
+  double evt_BSy0_ = spotEvent.y0();
+  double evt_BSz0_ = spotEvent.z0();
+  double evt_Beamsigmaz_ = spotEvent.sigmaZ();
+  double evt_BeamWidthX_ = spotEvent.BeamWidthX();
+  double evt_BeamWidthY_ = spotEvent.BeamWidthY();
+
+  if (!dbFromEvent_) {
+    edm::ESHandle<BeamSpotObjects> beamhandle = iSetup.getHandle(bsFromDBToken_);
+    const BeamSpotObjects* aSpot = beamhandle.product();
+
+    // translate from BeamSpotObjects to reco::BeamSpot
+    reco::BeamSpot::Point apoint(aSpot->x(), aSpot->y(), aSpot->z());
+
+    reco::BeamSpot::CovarianceMatrix matrix;
+    for (int i = 0; i < 7; ++i) {
+      for (int j = 0; j < 7; ++j) {
+        matrix(i, j) = aSpot->covariance(i, j);
```

Comment:
curiously, the ES BS covariance is not symmetric by construction `covariance_[7][7]`; seems wasteful and also dangerous for numerical precision; sigh

Reply:
Is there any action item from this comment?